allocator: Fix panic when allocations happen at init time

aaronlehmann opened the pull request:
`a.netCtx` is initialized too late, so if allocations happen as part of
`doNetworkInit`, a nil pointer dereference will cause a panic.


Initialize `a.netCtx` earlier and use `a.netCtx` directly in member
functions instead of passing the network context separately, so there is
no confusion about which to use.

Also change allocator.go to have separate entries in the waitgroup for
initialization and actually running the allocator, and defer `Done` for
both. This should prevent a panic like this from leading to a deadlock,
since the deferred `Done` will be reached.

See docker/docker#25432

cc @mrjana @LK4D4 @tonistiigi

aaronlehmann commented:
I think we should also make sure this scenario is covered in unit tests. I'm not very familiar with the tests for this part of the code, so it would be great if someone could point me in the right direction or submit a separate PR for test coverage.

LK4D4 commented:
LGTM

mrjana commented on manager/allocator/network.go:
```diff
@@ -81,6 +81,7 @@ func (a *Allocator) doNetworkInit(ctx context.Context) error {
 		unallocatedNetworks: make(map[string]*api.Network),
 		ingressNetwork:      newIngressNetwork(),
 	}
+	a.netCtx = nc
```
The reason `a.netCtx` is initialized at the end is to make sure in case of failures we don't return from `doNetworkInit` with a `netCtx` which shouldn't be there in the `Allocator`. This is because the `Allocator` has a longer life time than what happens in `doNetworkInit` itself. 

So in all of `doNetworkInit` I've just passed the `netCtx` as an argument to functions that need it. May be we should do the same for `taskCreateNetworkAttachments` i.e add `netCtx` as an argument?

aaronlehmann replied:
Sounds kind of dangerous TBH. What about a deferred closure in `doNetworkInit` that clears `a.netCtx` if an error is being returned?

aaronlehmann commented:
@mrjana: Updated to add a closure that clears `a.netCtx` in the error case, and removed `networkContext` parameters from the functions so there's no confusion over whether to use the argument or `a.netCtx`.

mrjana commented:
@aaronlehmann Adding a test case to cover this scenario would be nice. I will try to add one.

aluzzardi commented:
@aaronlehmann @mrjana Please add a test case in a separate PR
